zproject/backends.py

```diff
@@ -550,11 +563,91 @@ def get_authenticated_user(self, *args: Any, **kwargs: Any) -> Optional[UserProf
 
         return user_profile
 
+class SAMLAuthBackend(SocialAuthMixin, SAMLAuth):
+    auth_backend_name = "SAML"
+
+    def auth_complete(self, *args, **kwargs):
+        """
+        The user has been redirected back from the IdP and we should
+        now log them in, if everything checks out.
+        """
+        # This is copy-pasted from the SAML backend to use our own do_auth
+        # instead of self.strategy.authenticate
+        idp_name = self.strategy.request_data()['RelayState']
+        idp = self.get_idp(idp_name)
+        auth = self._create_saml_auth(idp)
+        auth.process_response()
+        errors = auth.get_errors()
+        if errors or not auth.is_authenticated():
+            reason = auth.get_last_error_reason()
+            raise AuthFailed(
+                self, 'SAML login failed: {0} ({1})'.format(errors, reason)
+            )
+
+        attributes = auth.get_attributes()
+        attributes['name_id'] = auth.get_nameid()
+        self._check_entitlements(idp, attributes)
+        response = {
+            'idp_name': idp_name,
+            'attributes': attributes,
+            'session_index': auth.get_session_index(),
+        }
+        kwargs.update({'response': response, 'backend': self})
+        return self.do_auth(*args, **kwargs)
```

Comment:
Now that `35c4a9f1d2b3adf5820cc239afd5caba2062b93f` is merged, I think you shouldn't need this function at all.  Basically, our code will run after SAML auth is finished via the `PIPELINE` declarations in `zproject/settings.py`.